lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -681,9 +684,37 @@ def _push_alerts_to_relation_databag(self, relation: Relation) -> None:
 
         alert_rules_as_dict = alert_rules.as_dict()
 
+        alert_rules_as_dict = self._parse_and_add_extra_alerts(alert_rules_as_dict)
+
         if alert_rules_as_dict:
             relation.data[self._charm.app]["alert_rules"] = json.dumps(alert_rules_as_dict)
 
+    def _parse_and_add_extra_alerts(self, alert_rules_as_dict: dict) -> dict:
+        """Parse alerts from `self.extra_alerts_callable`.
+
+        Call `self.extra_alerts_callable`, parse alert groups and
+        extend (mutate) `alert_rules_as_dict` with new alerts.
+
+        Returns:
+            (dict) Modified `alert_rules_as_dict`. Return is done only for explicit mutation.
+        """
+        if callable(self.extra_alerts_callable):
+            extra_alerts_list = []
+            extra_alerts = self.extra_alerts_callable()
+            if extra_alerts:
+                for topology_identifier, alert_rule_groups in extra_alerts.items():
+                    extra_alerts_list.extend(alert_rule_groups.get("groups", []))
+
+                if alert_rules_as_dict:
+                    alert_rules_as_dict["groups"].extend(extra_alerts_list)
```

Comment:
This is an in-place modification ("by-ref"). Should probably make a copy first just in case? Or make it clear it's modifying in-place (i.e. do not return anything)?

Reply:
I don't know why we'd make a copy "just in case" which we'll just garbage collect later. It's not performance sensitive, but an extra stack allocation for no reason probably isn't worth it.

On the other hand, I would still say that we could just move this up and take:

There are a couple of ways in which all of this could be rearranged to be slightly cleaner, but I'm not going to stand on ceremony about nits.

Reply:
function is refactored to be cleaner. 

However, without a copy. Importing deepcopy, etc does not make a value here. It is not a loop but one time extension, so we do not reference mutable item several times